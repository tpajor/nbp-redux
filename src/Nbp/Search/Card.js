import React from 'react';
import PropTypes from 'prop-types';

import './Card.css';

const Card = ({ cardData, isTemporary, deleteCard, userSignedIn, signedIn, addCard, showDetails, cards }) => {
  return (
    <div className={'Card'}>
      <div className={'row CardMainContainer'}>
        <div className={'col-4 CardMainCell Left'}>
          <p className={'CardMainCellText'}>
            {cardData.code}
          </p>
        </div>
        <div className={'col-4 CardMainCell Mid'}>
          <p className={'CardMainCellText'}>
            {cardData.rates[0].ask}  
          </p>  
        </div>
        <div className={'col-4 CardMainCell Right'}>
          <p className={'CardMainCellText'}>
            {cardData.rates[0].bid}
          </p>
          {isTemporary ? 
            '' :
            <button className={'DeleteCardButton'} onClick={() => deleteCard(cardData.id, userSignedIn, signedIn)}>X</button>
          }
        </div>
      </div>
      <div className={'CardNameCell'}>
        <p>
          {cardData.currency}
        </p>
      </div>
      { (isTemporary) ?
        <div className={'DoubleButton'}>
          <button className={'SaveButton'} onClick={() => addCard(cardData, userSignedIn, signedIn, cards)}>
            Przypnij
          </button> 
          <button className={'HalfCardButton'} onClick={() => showDetails(cardData.code)}>
            Dodaj do wykresu
          </button>
        </div> :
        <button className={'CardButton'} onClick={() => showDetails(cardData.code)}>
          Dodaj do wykresu
        </button>
      }
    </div>
  );
};

Card.propTypes = {
  userSignedIn: PropTypes.string,
  signedIn: PropTypes.bool,
  cardData: PropTypes.object,
  addCard: PropTypes.func,
  showDetails: PropTypes.func,
  isTemporary: PropTypes.bool,
  deleteCard: PropTypes.func,
  cards: PropTypes.array,
};

export default Card;