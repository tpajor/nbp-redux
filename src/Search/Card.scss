.Card {
  height: 150px;
  width: 70%;
  margin: 20px 20% 0 10%;
  border: 1px solid #ccc;
  border-radius: 5px;
  box-shadow: 5px 5px #aaa;
}

.CardMainContainer {
  margin: 0 auto;
  width: 100%;
  height: 60%;
}

.CardMainCell {
  height: 100%;
  border: 1px solid #ccc;
  border-style: none solid solid none;
  text-align: center;
}

.Left {
  border-style: none solid solid none;
}

.Mid {
  border-style: none none solid none;
}

.Right {
  border-style: none none solid solid;
  position: relative;
}

.CardMainCellText {
  font-size: 2em;
  margin-top: 15%;
}

.DeleteCardButton {
  position: absolute;
  border: none;
  background-color: white;
  right: 1px;
  top: 1px;
  cursor: pointer;
  color: gray;
}

.CardNameCell {
  height: 20%;
  width: 100%;
  text-align: center;
}

.CardButton {
  width: 100%;
  height: 20%;
  background-color: #ccc;
  cursor: pointer;
  border: none;
}

.SaveButton {
  border: 1px solid black;
  border-style: none solid none none;
  width: 50%;
  height: 100%;
  background-color: #ccc;
  cursor: pointer;
}

.DoubleButton {
  height: 20%;
}

.HalfCardButton {
  width: 50%;
  height: 100%;
  background-color: #ccc;
  cursor: pointer;
  border: none;
}