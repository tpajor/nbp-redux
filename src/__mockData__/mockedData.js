const card = { id: 'someId', code: 'USD', currency: 'dolar amerykański', rates: [{ ask: 3.3, bid: 3.4 }]};
const card2 = { id: 'someId2', code: 'USD', currency: 'dolar amerykański', rates: { ask: 3.3, bid: 3.4 }};

const search = {
  cards: [card, card2],
  temporaryCard: card,
  searchError: false,
  addCardError: true,
  addCardErrorMessage: 'error',
  currenciesTable: ['usd', 'chf', 'aud'],
  currencyCodeToViewInDetail: 'usd',
};

const user = { userName: 'User', pass: 'pass' };
const user2 = { userName: 'User2', pass: 'pass2'};

export const login = {
  signingError: true,
  signingErrorMessage: 'error',
  signedIn: true,
  userSignedIn: 'User',
  users: [user, user2]
};

const lastRates = [
  { ask: 3.5983, bid: 3.5271, effectiveDate: "2018-05-07", no: "087/C/NBP/2018" },
  { ask: 3.6088, bid: 3.5374, effectiveDate: "2018-05-08", no: "088/C/NBP/2018" }
];

const detail = {
  lastRates: lastRates,
  detailsInputError: false,
  detailsInputErrorMessage: '',
};

export const initialState = { search, login, detail };

export const searchViewData = {
  cards: search.cards,
  temporaryCard: search.temporaryCard,
  searchError: search.searchError,
  signedIn: login.signedIn,
  userSignedIn: login.userSignedIn,
  addCardError: search.addCardError,
  addCardErrorMessage: search.addCardErrorMessage,
  currenciesTable: search.currenciesTable,
};

export const detailViewData = {
  currencyCodeToViewInDetail: search.currencyCodeToViewInDetail,
  lastRates: detail.lastRates,
  detailsInputError: detail.detailsInputError,
  detailsInputErrorMessage: detail.detailsInputErrorMessage,
  signedIn: login.signedIn,
};