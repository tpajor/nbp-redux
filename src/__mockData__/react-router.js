const Link = () => {
  return jest.fn();
}

export default Link;