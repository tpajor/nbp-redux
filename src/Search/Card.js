import React from 'react';
import PropTypes from 'prop-types';

import './Card.css';

const Card = (props) => {
  return (
    <div className={'Card'}>
      <div className={'row CardMainContainer'}>
        <div className={'col-4 CardMainCell Left'}>
          <p className={'CardMainCellText'}>
            {props.cardData.code}
          </p>
        </div>
        <div className={'col-4 CardMainCell Mid'}>
          <p className={'CardMainCellText'}>
            {props.cardData.rates[0].ask}  
          </p>  
        </div>
        <div className={'col-4 CardMainCell Right'}>
          <p className={'CardMainCellText'}>
          {props.cardData.rates[0].bid}
          </p>
          {props.isTemporary ? 
            '' :
            <button className={'DeleteCardButton'} onClick={() => props.deleteCard(props.cardData.id)}>X</button>
          }
        </div>
      </div>
      <div className={'CardNameCell'}>
        <p>
          {props.cardData.currency}
        </p>
      </div>
      { (props.isTemporary) ?
        <div className={'DoubleButton'}>
          <button className={'SaveButton'} onClick={() => props.addCard(props.cardData)}>
            Zapisz
          </button> 
          <button className={'HalfCardButton'} onClick={() => props.showDetails(props.cardData.code)}>
            Szczegóły
          </button>
        </div> :
        <button className={'CardButton'} onClick={() => props.showDetails(props.cardData.code)}>
          Szczegóły
        </button>
      }
    </div>
  );
};

Card.propTypes = {
  cardData: PropTypes.object,
  addCard: PropTypes.func,
  showDetails: PropTypes.func,
  isTemporary: PropTypes.bool,
  deleteCard: PropTypes.func,
};

export default Card;