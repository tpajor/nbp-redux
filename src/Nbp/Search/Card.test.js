import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Card from './Card';
import { searchViewData } from '../../__mockData__/mockedData'


Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const props = {
    cardData: searchViewData.temporaryCard,
    isTemporary: true,
    userSignedIn: searchViewData.userSignedIn,
    signedIn: searchViewData.signedIn,
    cards: searchViewData.cards,
    addCard: jest.fn(),
    deleteCard: jest.fn(),
    showDetails: jest.fn(),
  }

  const enzymeWrapper = shallow(<Card {...props} />);

  return {
    enzymeWrapper,
    props
  };
};

describe('Card component', () => {
  it('should render self and subcomponents with their props', () => {
    const { enzymeWrapper, props } = setup();

    expect(enzymeWrapper.length).toBe(1);

    expect(enzymeWrapper.find('div').length).toBe(7);
    expect(enzymeWrapper.find('div').first().hasClass('Card')).toBe(true);
    expect(enzymeWrapper.find('div').at(1).hasClass('row CardMainContainer')).toBe(true);
    expect(enzymeWrapper.find('div').at(2).hasClass('col-4 CardMainCell Left')).toBe(true);
    expect(enzymeWrapper.find('div').at(3).hasClass('col-4 CardMainCell Mid')).toBe(true);
    expect(enzymeWrapper.find('div').at(4).hasClass('col-4 CardMainCell Right')).toBe(true);
    expect(enzymeWrapper.find('div').at(5).hasClass('CardNameCell')).toBe(true);
    expect(enzymeWrapper.find('div').last().hasClass('DoubleButton')).toBe(true);

    expect(enzymeWrapper.find('p').length).toBe(4);
    expect(enzymeWrapper.find('p').first().hasClass('CardMainCellText')).toBe(true);
    expect(enzymeWrapper.find('p').at(1).hasClass('CardMainCellText')).toBe(true);
    expect(enzymeWrapper.find('p').at(2).hasClass('CardMainCellText')).toBe(true);

    expect(enzymeWrapper.find('button').length).toBe(2);
    expect(enzymeWrapper.find('button').first().hasClass('SaveButton')).toBe(true);
    expect(enzymeWrapper.find('button').last().hasClass('HalfCardButton')).toBe(true);
    enzymeWrapper.setProps({ ...props, isTemporary: false });
    expect(enzymeWrapper.find('button').length).toBe(2);
    expect(enzymeWrapper.find('button').first().hasClass('DeleteCardButton')).toBe(true);
    expect(enzymeWrapper.find('button').last().hasClass('CardButton')).toBe(true);
    enzymeWrapper.setProps({ ...props });
  });

  it('should render correctly', () => {
    const { enzymeWrapper } = setup();
    
    expect(enzymeWrapper).toMatchSnapshot();
  });

  it('should call deleteCard props function when the right buuton is clicked', () => {
    const { enzymeWrapper, props } = setup();
    enzymeWrapper.setProps({ ...props, isTemporary: false })
    const deleteButton = enzymeWrapper.find('button').first();
    deleteButton.simulate('click');
    expect(props.deleteCard.mock.calls.length).toBe(1);
  });

  it('should call showDetails props function when the right buuton is clicked', () => {
    const { enzymeWrapper, props } = setup();
    let showDetailsButton = enzymeWrapper.find('button').last();
    showDetailsButton.simulate('click');
    expect(props.showDetails.mock.calls.length).toBe(1);

    enzymeWrapper.setProps({ ...props, isTemporary: false })
    showDetailsButton = enzymeWrapper.find('button').last();
    showDetailsButton.simulate('click');
    expect(props.showDetails.mock.calls.length).toBe(2);
  });

  it('should call addCard props function when the right buuton is clicked', () => {
    const { enzymeWrapper, props } = setup();
    const addButton = enzymeWrapper.find('button').first();
    addButton.simulate('click');
    expect(props.addCard.mock.calls.length).toBe(1);
  });
});
