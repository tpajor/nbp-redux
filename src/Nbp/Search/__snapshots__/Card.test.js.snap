// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    addCard={[Function]}
    cardData={
        Object {
            "code": "USD",
            "currency": "dolar amerykański",
            "id": "someId",
            "rates": Array [
              Object {
                "ask": 3.3,
                "bid": 3.4,
              },
            ],
          }
    }
    cards={
        Array [
            Object {
              "code": "USD",
              "currency": "dolar amerykański",
              "id": "someId",
              "rates": Array [
                Object {
                  "ask": 3.3,
                  "bid": 3.4,
                },
              ],
            },
            Object {
              "code": "USD",
              "currency": "dolar amerykański",
              "id": "someId2",
              "rates": Object {
                "ask": 3.3,
                "bid": 3.4,
              },
            },
          ]
    }
    deleteCard={[Function]}
    isTemporary={true}
    showDetails={[Function]}
    signedIn={true}
    userSignedIn="User"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="row CardMainContainer"
>
          <div
                    className="col-4 CardMainCell Left"
          >
                    <p
                              className="CardMainCellText"
                    >
                              USD
                    </p>
          </div>
          <div
                    className="col-4 CardMainCell Mid"
          >
                    <p
                              className="CardMainCellText"
                    >
                              3.3
                    </p>
          </div>
          <div
                    className="col-4 CardMainCell Right"
          >
                    <p
                              className="CardMainCellText"
                    >
                              3.4
                    </p>
                    
          </div>
</div>,
        <div
          className="CardNameCell"
>
          <p>
                    dolar amerykański
          </p>
</div>,
        <div
          className="DoubleButton"
>
          <button
                    className="SaveButton"
                    onClick={[Function]}
          >
                    Przypnij
          </button>
          <button
                    className="HalfCardButton"
                    onClick={[Function]}
          >
                    Dodaj do wykresu
          </button>
</div>,
      ],
      "className": "Card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-4 CardMainCell Left"
>
              <p
                            className="CardMainCellText"
              >
                            USD
              </p>
</div>,
            <div
              className="col-4 CardMainCell Mid"
>
              <p
                            className="CardMainCellText"
              >
                            3.3
              </p>
</div>,
            <div
              className="col-4 CardMainCell Right"
>
              <p
                            className="CardMainCellText"
              >
                            3.4
              </p>
              
</div>,
          ],
          "className": "row CardMainContainer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p
                className="CardMainCellText"
>
                USD
</p>,
              "className": "col-4 CardMainCell Left",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "USD",
                "className": "CardMainCellText",
              },
              "ref": null,
              "rendered": "USD",
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <p
                className="CardMainCellText"
>
                3.3
</p>,
              "className": "col-4 CardMainCell Mid",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 3.3,
                "className": "CardMainCellText",
              },
              "ref": null,
              "rendered": 3.3,
              "type": "p",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="CardMainCellText"
>
                  3.4
</p>,
                "",
              ],
              "className": "col-4 CardMainCell Right",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 3.4,
                  "className": "CardMainCellText",
                },
                "ref": null,
                "rendered": 3.4,
                "type": "p",
              },
              "",
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <p>
            dolar amerykański
</p>,
          "className": "CardNameCell",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "dolar amerykański",
          },
          "ref": null,
          "rendered": "dolar amerykański",
          "type": "p",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="SaveButton"
              onClick={[Function]}
>
              Przypnij
</button>,
            <button
              className="HalfCardButton"
              onClick={[Function]}
>
              Dodaj do wykresu
</button>,
          ],
          "className": "DoubleButton",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Przypnij",
              "className": "SaveButton",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Przypnij",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Dodaj do wykresu",
              "className": "HalfCardButton",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Dodaj do wykresu",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row CardMainContainer"
>
            <div
                        className="col-4 CardMainCell Left"
            >
                        <p
                                    className="CardMainCellText"
                        >
                                    USD
                        </p>
            </div>
            <div
                        className="col-4 CardMainCell Mid"
            >
                        <p
                                    className="CardMainCellText"
                        >
                                    3.3
                        </p>
            </div>
            <div
                        className="col-4 CardMainCell Right"
            >
                        <p
                                    className="CardMainCellText"
                        >
                                    3.4
                        </p>
                        
            </div>
</div>,
          <div
            className="CardNameCell"
>
            <p>
                        dolar amerykański
            </p>
</div>,
          <div
            className="DoubleButton"
>
            <button
                        className="SaveButton"
                        onClick={[Function]}
            >
                        Przypnij
            </button>
            <button
                        className="HalfCardButton"
                        onClick={[Function]}
            >
                        Dodaj do wykresu
            </button>
</div>,
        ],
        "className": "Card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col-4 CardMainCell Left"
>
                <p
                                className="CardMainCellText"
                >
                                USD
                </p>
</div>,
              <div
                className="col-4 CardMainCell Mid"
>
                <p
                                className="CardMainCellText"
                >
                                3.3
                </p>
</div>,
              <div
                className="col-4 CardMainCell Right"
>
                <p
                                className="CardMainCellText"
                >
                                3.4
                </p>
                
</div>,
            ],
            "className": "row CardMainContainer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p
                  className="CardMainCellText"
>
                  USD
</p>,
                "className": "col-4 CardMainCell Left",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "USD",
                  "className": "CardMainCellText",
                },
                "ref": null,
                "rendered": "USD",
                "type": "p",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <p
                  className="CardMainCellText"
>
                  3.3
</p>,
                "className": "col-4 CardMainCell Mid",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": 3.3,
                  "className": "CardMainCellText",
                },
                "ref": null,
                "rendered": 3.3,
                "type": "p",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="CardMainCellText"
>
                    3.4
</p>,
                  "",
                ],
                "className": "col-4 CardMainCell Right",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": 3.4,
                    "className": "CardMainCellText",
                  },
                  "ref": null,
                  "rendered": 3.4,
                  "type": "p",
                },
                "",
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <p>
              dolar amerykański
</p>,
            "className": "CardNameCell",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "dolar amerykański",
            },
            "ref": null,
            "rendered": "dolar amerykański",
            "type": "p",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="SaveButton"
                onClick={[Function]}
>
                Przypnij
</button>,
              <button
                className="HalfCardButton"
                onClick={[Function]}
>
                Dodaj do wykresu
</button>,
            ],
            "className": "DoubleButton",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Przypnij",
                "className": "SaveButton",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Przypnij",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Dodaj do wykresu",
                "className": "HalfCardButton",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Dodaj do wykresu",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
